blogs/treewidth.md

```diff
@@ -0,0 +1,385 @@
+
+A formal definition of the [**line graph**](https://en.wikipedia.org/wiki/Line_graph) is as follows:
+
+**Definition 1**. Given a graph G, its line graph $L(G)$ is a graph such that: 
+   1. each vertex of $L(G)$ represents an edge of $G$; 
+   2. two vertices of $L(G)$ are adjacent if and only if their corresponding edges share a common endpoint ("are incident") in $G$. 
+
+That is, it is the intersection graph of the edges of G, representing each edge by the set of its two endpoints.
+
+For a tensor network, we can construct a hypergraph $G$ whose vertices are the tensors and whose hyperedges are the indices. Then the line graph $L(G)$ of the hypergraph $G$ is the graph whose vertices are the indices and whose edges are the tensors, which is a simple graph, as shown in the following figure:
+
+![Fig.1](/assets/treewidth_figs/linegraph.png)
+
+Since we are considering a tensor network, dimension of the indices have to be considered. 
+Therefore, for each vertex of the line graph $L(G)$, we define its weight as $\log_2(d)$, where $d$ is the dimension of the index. 
+
+In this way, any possible tensor can be represented as a subgraph induced by a set of vertices in $L(G)$, and its log2 size is given by the sum of weights of the boundary vertices of the subgraph. For example, the tensor $C$ in the figure above corresponding to the subgraph induced by the vertices $\{b, d, f\}$, and its log2 size is 
+$${dims(b) \times dims(d) \times dims(f)} = 2^{w(b) + w(d) + w(f)}$$
+where $w(*)$ for weights of vertices.
+
+### Tree Decomposition and Tree Width
+
+Intuitively, a [**tree decomposition**](https://en.wikipedia.org/wiki/Tree_decomposition) represents the vertices of a given graph $G$ as subtrees of a tree, in such a way that vertices in $G$ are adjacent only when the corresponding subtrees intersect.
+
+The tree decomposition of a graph is a tree whose nodes are subsets of the vertices of the graph, and the following conditions are satisfied:
+
+1. Each vertex of the graph is in at least one node of the tree.
+2. For each edge of the graph, there is a node of the tree containing both vertices of the edge.
+3. Bags containing the same vertex have to be connected in the tree.
+
+An example of tree decomposition is shown in the following figure:
+
+![](/assets/treewidth_figs/treedecomposition.png)
+
+where the left graph is the original graph and the right one is the tree decomposition of the graph.
+
+All the nodes of the tree are called **tree bags**, and the intersection of two bags is called a **separator**.
+The width of a tree decomposition is the size of the largest bag minus one, thus the width of the example above is $2$.
+Clearly, one graph can have multiple tree decompositions, and the **tree width** of a graph is the minimum width of all possible tree decompositions.
+For graphs with weights on vertices, we define the width as sum of weights of the vertices in the bags minus one.
+
+### From Tree Decomposition to Contraction Order
+
```

Comment:
In this section, I failed to understand how the example actually relates to the tree decomposition of the line graph, and thus to properly see the connection between the minimal three width and tree bag size on the one hand, and the computational and space complexity. I feel like this is one of the most important sections of the blog, but I failed to fully understand it.

Reply:
I am sorry for the unclear statement in the previous version. In the revised version, I tried to give a clearer statement about the relation between tree decomposition and the contraction order in the appendix, and we added more example to explain that.